internal/controller/controllers/installenv_controller.go

```diff
@@ -0,0 +1,221 @@
+	} else {
+		isoParams := installer.GenerateClusterISOParams{
+			ClusterID:         *c.ID,
+			ImageCreateParams: &models.ImageCreateParams{},
+		}
+		if len(installEnv.Spec.SSHAuthorizedKeys) > 0 {
+			isoParams.ImageCreateParams.SSHPublicKey = installEnv.Spec.SSHAuthorizedKeys[0]
+		}
+		// GenerateClusterISOInternal will generate an ISO only if there it was not generated before,
+		// or something has changed in isoParams.
+		updatedCluster, inventoryErr = r.Installer.GenerateClusterISOInternal(ctx, isoParams)
+	}
+
+	if inventoryErr == nil {
+		imageInfo = updatedCluster.ImageInfo
+		imageState.Status = corev1.ConditionTrue
+		imageState.State = ImageCreated
+		imageState.StateInfo = adiiov1alpha1.ImageStateCreated
+	}
+
+	return r.updateStatusAndReturnResult(ctx, installEnv, imageInfo, imageState, inventoryErr)
+}
+
+func (r *InstallEnvReconciler) updateStatusAndReturnResult(
+	ctx context.Context,
+	installEnv *adiiov1alpha1.InstallEnv,
+	imageInfo *models.ImageInfo,
+	imageState ImageState,
+	err error) (ctrl.Result, error) {
+
+	var res ctrl.Result
+
+	if imageBeingCreated(err) {
+		imageState.Status = corev1.ConditionTrue
+		imageState.State = ImageCreated
+		imageState.StateInfo = adiiov1alpha1.ImageStateCreated
+		// Clear up the error stateInfo while installEnv is being created
+		err = nil
+		r.Log.Infof("Image %s being prepared for cluster %s stateInfo: %s",
+			installEnv.Name, installEnv.ClusterName, imageState.StateInfo)
+	} else if isClientError(err) {
+		imageState.Status = corev1.ConditionFalse
+		imageState.State = ImageCreationError
+		imageState.StateInfo += ": " + err.Error()
+	} else if err != nil {
+		imageState.Status = corev1.ConditionFalse
+		imageState.State = ImageCreationError
+		imageState.StateInfo += ": internal error"
+		res.Requeue = true
+	}
+	r.setStateAndStateInfo(adiiov1alpha1.ImageCreatedCondition, imageState, installEnv)
+
+	if err == nil && imageInfo != nil {
+		installEnv.Status.ISODownloadURL = imageInfo.DownloadURL
+	} else if err != nil {
+		r.Log.WithError(err).Error("installEnv reconcile failed")
+	}
+
+	if updateErr := r.Status().Update(ctx, installEnv); updateErr != nil {
+		r.Log.WithError(updateErr).Error("failed to update installEnv status")
+		res.Requeue = true
+		return res, nil
+	}
+
+	return res, nil
+}
+
+func (r *InstallEnvReconciler) setStateAndStateInfo(conditionType adiiov1alpha1.InstallEnvConditionType,
+	imageState ImageState, installEnv *adiiov1alpha1.InstallEnv) {
+	r.setCondition(adiiov1alpha1.InstallEnvCondition{
+		Type:               conditionType,
+		Status:             imageState.Status,
+		LastProbeTime:      metav1.Time{Time: time.Now()},
+		LastTransitionTime: metav1.Time{Time: time.Now()},
+		Reason:             imageState.State,
+		Message:            imageState.StateInfo,
+	}, &installEnv.Status.Conditions)
+}
+
+func (r *InstallEnvReconciler) setCondition(condition adiiov1alpha1.InstallEnvCondition, conditions *[]adiiov1alpha1.InstallEnvCondition) {
```

Comment:
Not sure if it helps you, but it wasn't too long ago that I wrote this library https://pkg.go.dev/github.com/openshift/custom-resource-status@v0.0.0-20200602122900-c002fd1547ca/conditions/v1 to make setting conditions easier.

Reply:
Even if you don't go the route of using the `custom-resource-status` library, it may be a good idea to put this logic (as well as `findInstallEnvConditionIndexByType`) in `internal/controller/api/v1alpha1/` and make it generic such that all controllers in this repo don't need to re-implement the logic when they decide to add conditions.

Reply:
Done. now using `openshift/custom-resource-status`